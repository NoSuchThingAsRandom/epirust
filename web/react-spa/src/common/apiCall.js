import config from "../config";

export function post(url, data) {
  return fetch(`${config.API_HOST}${url}`, {
    method: 'POST',
    headers: {'Content-Type': 'application/json'},
    body: JSON.stringify(data)
  })
    .then((response) => {
      if (!response.ok) {
        throw Error(response.statusText);
      }
      return response;
    })
}


