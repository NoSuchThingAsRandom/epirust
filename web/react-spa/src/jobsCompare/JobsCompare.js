/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import React, {useEffect, useState} from "react";
import ComparerDropdowns from "./ComparerDropdowns";
import {get} from "../common/apiCall";
import {LoadingComponent} from "../common/LoadingComponent";
import {LOADING_STATES} from "../common/constants";
import Graph from "../time-series/LineGraph";
import GraphUpdater from "./GraphUpdater";
import {reduceStatus} from "../jobs/JobTransformer";

export default function JobsCompare() {
  const [jobs, updateJobs] = useState([])
  const [loadingState, updateLoadingState] = useState(LOADING_STATES.LOADING)
  const [graphData, updateGraphData] = useState([])

  useEffect(() => {
    get('/jobs')
      .then((res) => res.json())
      .then((jobsResponse) => {
        updateJobs(jobsResponse.map(reduceStatus).filter(j => j.status === 'finished'))
        updateLoadingState(LOADING_STATES.FINISHED)
      })
      .catch(() => updateLoadingState(LOADING_STATES.FAILED))
  }, [])

  const updateBuffer = (buffer) => {
    updateGraphData(prevData => prevData.concat(buffer))
  }

  const onCompare = (selectedJobs) => {
    updateGraphData([])
    new GraphUpdater(updateBuffer, selectedJobs.job1, selectedJobs.job2).start()
  }

  return <div className='jobs-compare'>
    <LoadingComponent loadingState={loadingState}>
      <ComparerDropdowns jobs={jobs} onCompare={onCompare}/>
      <div className='jobs-compare-chart'>
        {graphData.length > 0 && <Graph dataBuffer={makeCSV(graphData)}/>}
      </div>
    </LoadingComponent>
  </div>
}

const rowToCsv = (row) => {
  if (row) return [row.susceptible, row.susceptible_std, row.infected, row.infected_std, row.quarantined,
    row.quarantined_std, row.recovered, row.recovered_std, row.deceased, row.deceased_std]
  return ['', 0, '', 0, '', 0, '', 0, '', 0]
}

const makeCSV = (graphData) => {

  return [["hour",
    "susceptible_job1", "infected_job1", "quarantined_job1", "recovered_job1", "deceased_job1",
    "susceptible_job2", "infected_job2", "quarantined_job2", "recovered_job2", "deceased_job2"
  ]].concat(graphData.map(d => ([d.hour, ...rowToCsv(d.job1), ...rowToCsv(d.job2)]))).join('\n')
}
