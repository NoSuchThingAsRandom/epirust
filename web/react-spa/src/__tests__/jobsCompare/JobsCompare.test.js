/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import React from "react";
import ShallowRenderer from 'react-test-renderer/shallow';
import {render} from '@testing-library/react'
import JobsCompare from "../../jobsCompare/JobsCompare";
import {get} from '../../common/apiCall'
import {flushPromises} from "../helper/promiseHelper";
import {act} from "react-dom/test-utils";

jest.mock('../../common/apiCall')

describe('Jobs Compare', function () {
  it('should render with default state', function () {
    const renderer = new ShallowRenderer();
    renderer.render(<JobsCompare/>)

    expect(renderer.getRenderOutput()).toMatchSnapshot()
  });

  it('should render with jobs fetched', async function () {
    const jobs = [{_id: 1}, {_id: 2}];
    get.mockResolvedValueOnce({json: jest.fn().mockResolvedValueOnce(jobs)})
    const {container} = render(<JobsCompare/>)

    await act(async () => {
      await flushPromises()
    })

    expect(container).toMatchSnapshot()
  });

  it('should display failed message when fetching jobs has failed', async function () {
    get.mockRejectedValue()
    const {container} = render(<JobsCompare/>)

    await act(async () => {
      await flushPromises()
    })

    expect(container).toMatchSnapshot()
  });
});