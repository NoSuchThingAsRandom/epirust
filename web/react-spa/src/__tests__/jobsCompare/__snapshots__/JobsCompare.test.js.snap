// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Jobs Compare should display failed message when fetching jobs has failed 1`] = `
<div>
  <div>
    Failed to load
  </div>
</div>
`;

exports[`Jobs Compare should render with default state 1`] = `
<React.Fragment>
  <LoadingComponent
    loadingState={1}
  >
    <ComparerDropdowns
      jobs={Array []}
    />
  </LoadingComponent>
</React.Fragment>
`;

exports[`Jobs Compare should render with jobs fetched 1`] = `
<div>
  <div
    class="comparer-dropdowns container"
  >
    <form
      class="row justify-content-md-center"
    >
      <div
        class="col-4"
      >
        <label
          for="job1"
        >
          Job 1
        </label>
        <select
          class="form-control"
          id="job1"
          name="job1"
          required=""
        >
          <option
            disabled=""
            value=""
          >
             -- select a job -- 
          </option>
          <option>
            1
          </option>
          <option>
            2
          </option>
        </select>
      </div>
      <div
        class="col-4"
      >
        <label
          for="job2"
        >
          Job 2
        </label>
        <select
          class="form-control"
          id="job2"
          name="job2"
          required=""
        >
          <option
            disabled=""
            value=""
          >
             -- select a job -- 
          </option>
          <option>
            1
          </option>
          <option>
            2
          </option>
        </select>
      </div>
      <div
        class="col-2 margin-top-auto"
      >
        <button
          class="btn btn-primary"
        >
          Compare
        </button>
      </div>
    </form>
  </div>
</div>
`;
